Sources/Plasma/FeatureLib/pfMetalPipeline/plMetalEnumerate.mm

```diff
@@ -0,0 +1,92 @@
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+Additional permissions under GNU GPL version 3 section 7
+
+If you modify this Program, or any covered work, by linking or
+combining it with any of RAD Game Tools Bink SDK, Autodesk 3ds Max SDK,
+NVIDIA PhysX SDK, Microsoft DirectX SDK, OpenSSL library, Independent
+JPEG Group JPEG library, Microsoft Windows Media SDK, or Apple QuickTime SDK
+(or a modified version of those libraries),
+containing parts covered by the terms of the Bink SDK EULA, 3ds Max EULA,
+PhysX SDK EULA, DirectX SDK EULA, OpenSSL and SSLeay licenses, IJG
+JPEG Library README, Windows Media SDK EULA, or QuickTime SDK EULA, the
+licensors of this Program grant you additional
+permission to convey the resulting work. Corresponding Source for a
+non-source form of such a combination shall include the source code for
+the parts of OpenSSL and IJG JPEG Library used as well as that of the covered
+work.
+
+You can contact Cyan Worlds, Inc. by email legal@cyan.com
+ or by snail mail at:
+      Cyan Worlds, Inc.
+      14617 N Newport Hwy
+      Mead, WA   99021
+
+*==LICENSE==*/
+
+#include "HeadSpin.h"
+
+#include <vector>
+
+#include <Foundation/Foundation.h>
+
+#include "plMetalPipeline.h"
+#include <sys/utsname.h>
+
+void plMetalEnumerate::Enumerate(std::vector<hsG3DDeviceRecord>& records)
+{
+    //For now - just use the default device. If there is a high power discrete device - this will spin it up.
+    //This will also automatically pin us to an eGPU if present and the user has configured us to use it.
+    MTL::Device* device = MTL::CreateSystemDefaultDevice();
+    
+    if (device) {
+        hsG3DDeviceRecord devRec;
+        devRec.SetG3DDeviceType(hsG3DDeviceSelector::kDevTypeMetal);
+        devRec.SetDriverName("Metal");
+        devRec.SetDeviceDesc(device->name()->utf8String());
+        //Metal has ways to query capabilities, but doesn't expose a flat version
+        //Populate with the OS version
+        @autoreleasepool {
+            NSProcessInfo *processInfo = [NSProcessInfo processInfo];
+            NSOperatingSystemVersion version  = processInfo.operatingSystemVersion;
+            NSString *versionString = [NSString stringWithFormat:@"%li.%li.%li", (long)version.majorVersion, (long)version.minorVersion, version.patchVersion];
+            devRec.SetDriverVersion([versionString cStringUsingEncoding:NSUTF8StringEncoding]);
+        }
+        devRec.SetDriverDesc(device->name()->utf8String());
```

Comment:
Any chance we could use `hsSystemInfo::GetOperatingSystem()` here (and avoid contaminating our C++ code with Objective C)?

It would give slightly different output ("macOS 11.0" vs "11.0.0") but should still be reasonable?

Reply:
Yeah, I don't think that would be an issue. I don't think this string is used for anything critical?

I'm not sure Obj-C++ is the worst thing, but I understand the hesitation.

Reply:
Ooooof. `hsSystemInfo::GetOperatingSystem()` actually uses private API. I'd discourage its use until it can be fixed. That would be a blocker if the iOS App Store was ever a thing (which I know is a longer conversation.) In since there is a working implementation it might be better to stick with that for now. I can also look at rolling GetOperatingSystem() over to public API. That might require some Obj-C++ though.

Reply:
Regarding the private APIs, we discussed this a bit when `hsSystemInfo` was originally added: https://github.com/H-uru/Plasma/pull/1014#discussion_r748052130

Reply:
NSProcessInfo is the definitive way to do this under "modern" Cocoa. However, I'd still want to do some amount of digging before making a decision. IGetAppleVersion could be moved to Obj-C++.
https://developer.apple.com/documentation/foundation/nsprocessinfo/1410906-operatingsystemversion

Stuff like _kCFSystemVersionProductNameKey is actively being rejected by Apple. Not necessarily the highest priority in since this isn't going through the App Store (right now.) But I also wouldn't want to leave a risk like that lurking in the code for someone else to stumble upon.